src/extent.c

```diff
@@ -1587,8 +1587,23 @@ extent_record(tsdn_t *tsdn, arena_t *arena, extent_hooks_t **r_extent_hooks,
 	if (!extents->delay_coalesce) {
 		extent = extent_try_coalesce(tsdn, arena, r_extent_hooks,
 		    rtree_ctx, extents, extent, NULL, growing_retained);
+	} else if (extent_size_get(extent) >= LARGE_MINCLASS) {
+		/* Always coalesce large extents eagerly. */
+		bool coalesced;
+		size_t prev_size;
+		do {
+			prev_size = extent_size_get(extent);
+			assert(extent_state_get(extent) == extent_state_active);
+			extent = extent_try_coalesce(tsdn, arena,
+			    r_extent_hooks, rtree_ctx, extents, extent,
+			    &coalesced, growing_retained);
+			if (coalesced) {
+				extent_list_remove(&extents->lru, extent);
+				extent_state_set(extent, extent_state_active);
+			}
+		} while (coalesced &&
```

Comment:
Isn't this looping logic already done in extent_try_coalesce?

Reply:
When extents->delay_coalesce is true, each time extent_try_coalesce only coalesce one item. I thought about changing that (pass in a bool delay_coalesce instead of reading from extents), but then wanted to use the external control (stop trying coalescing if the last extent coalesced is small).